File: 2022/day02.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

int get_score(char a, char b){
    if (a == 'A'){ // rock
        if (b == 'X'){ // rock
            return 1 + 3;
        }
        else if (b == 'Y'){ // paper
            return 2 + 6;
        }
        else{ //scissors
            return 3;
        }
    }
    else if (a == 'B'){
        if (b == 'X'){ // rock
            return 1;
        }
        else if(b == 'Y'){// paper
            return 2 + 3;
        }
        else{ //scissors
            return 3 + 6;
        }
    }
    else{ //scissors
        if (b == 'X'){ // rock
            return  1 + 6;
        }
        else if (b == 'Y') {// paper
            return 2;
        }
        else{ // scissors
            return 3 + 3;
        }
    }
}

void first_part(){
    string line;
    ifstream inFile("input/day02.txt");
    int n = 0, total_score = 0;
    while(getline(inFile, line)) { 
        char a = line[0], b = line[2];
        total_score += get_score(a,b);
    }
    cout << total_score << endl; 
}

void second_part(){
    string line;
    ifstream inFile("input/day02.txt");
    int n = 0, total_score = 0;
    while(getline(inFile, line)) { 
        char a = line[0], result = line[2], b = ' ';

        if (a == 'A'){ // rock
            if (result == 'X'){ // lose
                b = 'Z';
            }
            else if (result == 'Y'){ // draw
                b = 'X';
            }
            else{ //win
                b = 'Y';
            }
        }
        else if (a == 'B'){
            if (result == 'X'){ // lose
                b = 'X';
            }
            else if(result == 'Y'){// draw
                b = 'Y';
            }
            else{ //win
                b = 'Z';
            }
        }
        else{ //scissors
            if (result == 'X'){ // lose
                b = 'Y';
            }
            else if (result == 'Y') {// draw
                b = 'Z';
            }
            else{ // win
                b = 'X';
            }
        }
        
        total_score += get_score(a,b);
    }
    cout << total_score << endl; 
}

int main() {
  // part 1
  first_part();

  // part 2
  second_part();
}

File: 2022/day02.py
def get_score(a, b):
    if a == 'A': # rock
        if b == 'X': # rock
            score = 1 + 3
        elif b == 'Y': # paper
            score = 2 + 6
        else: #scissors
            score = 3
    elif a == 'B': # paper
        if b == 'X': # rock
            score = 1
        elif b == 'Y': # paper
            score = 2 + 3
        else: #scissors
            score = 3 + 6
    else: #scissors
        if b == 'X': # rock
            score =  1 + 6

        elif b == 'Y': # paper
            score = 2

        else: #scissors
            score = 3 + 3
    return score

filename = "input/day02.txt"
# filename = "input/day02_sample.txt"

# part 1
with open(filename, 'r') as f:
    lines = f.readlines()
    total_score = 0
    for line in lines:
        a, b = line.split()
        score = get_score(a, b)
        total_score += score
print(total_score)

# part 2
with open(filename, 'r') as f:
    lines = f.readlines()
    total_score = 0
    for line in lines:
        a, result = line.split()
        
        b = ''
        if a == 'A': # rock
            if result == 'X': # lose
                b = 'Z'
            elif result == 'Y': # draw
                b = 'X'
            else: #win
                b = 'Y'
        elif a == 'B': # paper
            if result == 'X': # lose
                b = 'X'
            elif result == 'Y': # draw
                b = 'Y'
            else: # win
                b = 'Z'
        else: #scissors
            if result == 'X': # lose
                b = 'Y'
            elif result == 'Y': # draw
                b = 'Z'
            else: #win
                b = 'X'

        score = get_score(a, b)
        total_score += score
print(total_score)
